File: VB/CalDAVServer.SqlStorage.AspNet/Provisioning.vb
Imports System
Imports System.Configuration
Imports System.IO
Imports System.Web
Imports System.Security.Principal
Imports System.Security.AccessControl
Imports System.Threading.Tasks
Imports ITHit.WebDAV.Server.Acl
Imports CalDAVServer.SqlStorage.AspNet.CalDav

''' <summary>
''' This class creates initial calendar(s) and address book(s) for user during first log-in.
''' </summary>
Public Class Provisioning
    Implements IHttpModule

    Public Sub Dispose() Implements IHttpModule.Dispose
    End Sub

    Public Sub Init(application As HttpApplication) Implements IHttpModule.Init
        Dim postAuthAsyncHelper As EventHandlerTaskAsyncHelper = New EventHandlerTaskAsyncHelper(AddressOf App_OnPostAuthenticateRequestAsync)
        application.AddOnPostAuthenticateRequestAsync(postAuthAsyncHelper.BeginEventHandler, postAuthAsyncHelper.EndEventHandler)
    End Sub

    Private Async Function App_OnPostAuthenticateRequestAsync(source As Object, eventArgs As EventArgs) As Task
        Dim httpContext As HttpContext = HttpContext.Current
        If(httpContext.User Is Nothing) OrElse Not httpContext.User.Identity.IsAuthenticated Then Return
        Using context As DavContext = New DavContext(httpContext)
            ' Create calendars for the user during first log-in.
            Await CreateCalendarFoldersAsync(context)
            ' Closes transaction. Calls DavContextBaseAsync.BeforeResponseAsync only first time this method is invoked.
            ' This method must be called manually if DavContextBaseAsync is used outside of DavEngine. 
            Await context.EnsureBeforeResponseWasCalledAsync()
        End Using
    End Function

    ''' <summary>
    ''' Creates initial calendars for users.
    ''' </summary>
    Friend Shared Async Function CreateCalendarFoldersAsync(context As DavContext) As Task
        ' If user does not have access to any calendars - create new calendars.
        Dim sql As String = "SELECT ISNULL((SELECT TOP 1 1 FROM [cal_Access] WHERE [UserId] = @UserId) , 0)"
        If Await context.ExecuteScalarAsync(Of Integer)(sql, "@UserId", context.UserId) < 1 Then
            Await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 1", "Calendar 1")
            Await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 2", "Calendar 2")
            Await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 3", "Calendar 3")
        End If
    End Function
End Class

File: CS/CalDAVServer.SqlStorage.AspNet/Provisioning.cs
using System;
using System.Configuration;
using System.IO;
using System.Web;
using System.Security.Principal;
using System.Security.AccessControl;
using System.Threading.Tasks;

using ITHit.WebDAV.Server.Acl;
using CalDAVServer.SqlStorage.AspNet.CalDav;

namespace CalDAVServer.SqlStorage.AspNet
{
    /// <summary>
    /// This class creates initial calendar(s) and address book(s) for user during first log-in.
    /// </summary>
    public class Provisioning: IHttpModule
    {

        public void Dispose()
        {
        }

        public void Init(HttpApplication application)
        {
            EventHandlerTaskAsyncHelper postAuthAsyncHelper = new EventHandlerTaskAsyncHelper(App_OnPostAuthenticateRequestAsync);
            application.AddOnPostAuthenticateRequestAsync(postAuthAsyncHelper.BeginEventHandler, postAuthAsyncHelper.EndEventHandler);
        }

        private async Task App_OnPostAuthenticateRequestAsync(object source, EventArgs eventArgs)
        {
            HttpContext httpContext = HttpContext.Current;
            if ((httpContext.User == null) || !httpContext.User.Identity.IsAuthenticated)
                return;

            using (DavContext context = new DavContext(httpContext))
            {
                // Create calendars for the user during first log-in.
                await CreateCalendarFoldersAsync(context);

                // Closes transaction. Calls ContextAsync{IHierarchyItemAsync}.BeforeResponseAsync only first time this method is invoked.
                // This method must be called manually if ContextAsync{IHierarchyItemAsync} is used outside of DavEngine. 
                await context.EnsureBeforeResponseWasCalledAsync();
            }
        }

        /// <summary>
        /// Creates initial calendars for users.
        /// </summary>
        internal static async Task CreateCalendarFoldersAsync(DavContext context)
        {
            // If user does not have access to any calendars - create new calendars.
            string sql = @"SELECT ISNULL((SELECT TOP 1 1 FROM [cal_Access] WHERE [UserId] = @UserId) , 0)";
            if (await context.ExecuteScalarAsync<int>(sql, "@UserId", context.UserId) < 1)
            {
                await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 1", "Calendar 1");
                await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 2", "Calendar 2");
                await CalendarFolder.CreateCalendarFolderAsync(context, "Cal 3", "Calendar 3");
            }
        }
    }
}
